fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('http://localhost:3000/api/todos')
+      .then(res => res.json())
+      .then(data => this.setState({ todos: data }))
+      .catch(error => console.error(error));
+  }
+
+  /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+
+  addTodo(newTodo) {
+    fetch('http://localhost:3000/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(data => {
+        this.setState(state => ({ todos: [...state.todos, data] }));
+      })
+      .catch(error => {
+        console.error('Error:', error);
+      });
+  }
+
+  /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+
+  toggleCompleted(todoId) {
+    const item = this.state.todos.find(element => element.todoId === todoId);
+
+    fetch('http://localhost:3000/api/todos/' + todoId, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({ isCompleted: !item.isCompleted })
+    })
+      .then(response => response.json())
+      .then(data => {
+        const updatedList = this.state.todos.map(item => {
+          if (item.todoId === data.todoId) {
+            item.isCompleted = data.isCompleted;
+          }
```

Comment:
this is modifying the original object. instead of updating the object, just return data instead of item